### ISSUE TITLE
Memory pool for value lists

### ISSUE THREAD

**Issue by stoklund** [stoklund/cretonne#35]
The `InstructionData` struct is 16 bytes, and it is important to keep it small. Some instruction formats don't fit in 16 bytes, and so they use a `Box<FooData>` pointing to additional data allocated from the heap. Most notably, calls and branches keep their argument lists on the heap. This design has some disadvantages:

- Creating and destroying instructions causes malloc/free traffic which can be slow.
- Clearing the `DataFlowGraph` instruction list will fault in the whole array in order to call `Drop` code for the `InstructionData` members. This is slow.
- A `Vec<Value>` has 24 bytes of overhead before the first value is even added.
- A `Box<Vec<Value>>` leaves two pointer indirections between the instruction and the values.

These problems can be overcome by changing the large instruction representation to use references to value lists in a memory pool.

# Value list pool

Value lists are similar to `Vec<Value>`, but with a different layout:

- A `ValueList` is a 32-bit offset into the pool where the list is stored. Compare to the 24-byte footprint of a `Vec`.
- The length of the value list is stored next to the list in the pool array.
- The capacity is computed from the length, so it is not stored anywhere.

The pool is implemented as a single `Vec<Value>` shared by all the value lists used by a function. All manipulation of value lists goes through the pool:

```rust
pub struct ValueList(u32);
pub struct ValueListPool {...}
impl ValueListPool {
    fn alloc(&mut self, elements: &[Value]) -> ValueList;
    fn free(&mut self, vlist: ValueList);
    fn len(&self, vlist: ValueList) -> usize;
    fn get(&self, vlist: ValueList) -> &[Value];
    fn get_mut(&mut self, vlist: ValueList) -> &mut [Value];
    fn push(&mut self, vlist: &mut ValueList, element: Value);
    fn extend(&mut self, vlist: &mut ValueList, elements: &[Value]);
    fn insert(&mut self, vlist: &mut ValueList, index: usize, element: Value);
    fn remove(&mut self, vlist: &mut ValueList, index: usize);
}
```

The methods that change the length of a list are allowed to move it around, so they take a mutable reference to the list handle.

## Implementation

In order to get amortized linear time `push`, lists are reallocated exponentially. Capacities are arranged into power-of-two bins, computed from the length of the list:

```rust
fn bin_for_length(length: u32) -> u32 {
    30 - (length | 3).leading_zeros()
}

fn bin_size(bin: u32) -> u32 {
    4 << bin
}
```

The bin sizes are always powers of two, starting from 4. The capacity of a bin is its size minus one. The first word is used to store the list length.

When lists are freed, memory can be recycled by keeping a free list for each bin.

> Note that removing elements from a list may cause its capacity to shrink, and the list may have to be reallocated. This is not a problem, value lists are not used the same way as `Vec` in Cretonne.

# Instruction formats

The large instruction formats can now be reimplemented in terms of value lists:

- If an instruction format has too many fixed value arguments, or if it has `variable_args` arguments, move all value arguments to a `ValueList`.
- If an instruction format has both fixed and `variable_args` arguments, move both to a single value_list. This applies to `IndirectCall` only.
- All non-value arguments must fit in the `InstructionData` variant directly.

Changing the instruction formats like this has wide ranging effects:

- Remove the `boxed_storage` flag from the Python `InstructionFormat` class. Replace it with a `value_list` flag.
- Change the automatic instruction format mapping such that `(value, variable_args)` operands map to simply `(variable_args,)`, and do the same for instructions with too many value operands like `iadd_carry`.
- Change the generated `InstBuilder` methods to take `ValueList` arguments where needed.
- Change the IL parser.
- `InstructionData::args()` can be change to return a single slice, `&[Value]`, which is a nice simplification.

The `TernaryOverflow` instruction format becomes a catch-all for instructions with many inputs and many outputs and no immediate operands. Maybe rename it to `MultiAry` or something better.

**Comment by stoklund** [comment]
Implemented the `ValueList` in 0ada419fe7e as `EntityList<Value>`, they'll be needed by the register allocator too. Haven't updated the instruction representation yet.

**Comment by stoklund** [comment]
Fixed the instruction format part after some refactoring.
